docs/content/ref/registry.consul.register.deregisterCriticalServiceAfter.md

```diff
@@ -0,0 +1,11 @@
+---
+title: "registry.consul.register.deregisterCriticalServiceAfter"
+---
+
+`registry.consul.register.deregisterCriticalServiceAfter` configures the time the
+service health check is allowed to be in state `critical` until Consul automatically
+deregisters it.
```

Comment:
This has nothing to do with TTL check but only with the register feature as is in master. Meaning that if the HTTP check is `critical` after `registry.consul.register.deregisterCriticalServiceAfter` elapses, the service and all related checks are gone.

When introducing TTL with this PR, I didn't make it optional for the user to set the TTL check details and therefore think we don't need to make a reference to it here.

Reply:
Let me re-phrase. 

Assume that, for whatever reason, the HTTP check is critical long enough for Consul to remove the service and all its checks (which includes the TTL check). What happens when `c.Agent().UpdateTTL()` is called? Will this re-create the service or simply fail forever because the check doesn't exist anymore?

Reply:
The service will not be recreated by calling `c.Agent().UpdateTTL()`. At most, if this function is called after the service is removed - when could this happen? -, an error will take place because there's no such check - it was deleted with the service it was related to. Only if fabio re-register the service will the check (with the same check ID be present).

Reply:
With some more debugging, I was able to reproduce the scenario where the TTL expires without fabio exiting, and as can be observed, the service is re-registered ~90s after being created, meaning the TTL timed-out and Consul reaper deleted the service but fabio kicked-in to re-register it. This, I believe, is the expected behavior.

```
2019/06/14 09:18:32 [INFO] Version 1.5.11 starting
2019/06/14 09:18:32 [INFO] Go runtime is go1.12.5
2019/06/14 09:18:32 [INFO] Metrics disabled
2019/06/14 09:18:32 [INFO] Setting GOGC=800
2019/06/14 09:18:32 [INFO] Setting GOMAXPROCS=28
2019/06/14 09:18:32 [INFO] consul: Connecting to "localhost:8500" in datacenter "dc1"
2019/06/14 09:18:32 [INFO] Admin server access mode "rw"
2019/06/14 09:18:32 [INFO] Admin server listening on ":9998"
2019/06/14 09:18:32 [INFO] Waiting for first routing table
2019/06/14 09:18:32 [INFO] consul: Using dynamic routes
2019/06/14 09:18:32 [INFO] consul: Using tag prefix "urlprefix-"
2019/06/14 09:18:32 [INFO] consul: Watching KV path "/fabio/config"
2019/06/14 09:18:32 [INFO] consul: Watching KV path "/fabio/noroute.html"
2019/06/14 09:18:32 [INFO] HTTP proxy listening on :9999
2019/06/14 09:18:32 [INFO] Access logging disabled
2019/06/14 09:18:32 [INFO] Using routing strategy "rnd"
2019/06/14 09:18:32 [INFO] Using route matching "prefix"
2019/06/14 09:18:32 [INFO] consul: Registered fabio as "fabio"
2019/06/14 09:18:32 [INFO] consul: Registered fabio with id "fabio-pires-ipro.lan-9998"
2019/06/14 09:18:32 [INFO] consul: Registered fabio with address "192.168.86.187"
2019/06/14 09:18:32 [INFO] consul: Registered fabio with tags ""
2019/06/14 09:18:32 [INFO] consul: Registered fabio with check &{CheckID: Name: Args:[] DockerContainerID: Shell: Interval: Timeout: TTL:15s HTTP: Header:map[] Method: TCP: Status: Notes: TLSSkipVerify:false GRPC: GRPCUseTLS:false AliasNode: AliasService: DeregisterCriticalServiceAfter:15s}
2019/06/14 09:18:32 [INFO] consul: Registered fabio with check &{CheckID: Name: Args:[] DockerContainerID: Shell: Interval:1s Timeout:3s TTL: HTTP:http://[192.168.86.187]:9998/health Header:map[] Method: TCP: Status: Notes: TLSSkipVerify:false GRPC: GRPCUseTLS:false AliasNode: AliasService: DeregisterCriticalServiceAfter:90m}
2019/06/14 09:20:12 [INFO] consul: Registered fabio as "fabio"
2019/06/14 09:20:12 [INFO] consul: Registered fabio with id "fabio-pires-ipro.lan-9998"
2019/06/14 09:20:12 [INFO] consul: Registered fabio with address "192.168.86.187"
2019/06/14 09:20:12 [INFO] consul: Registered fabio with tags ""
2019/06/14 09:20:12 [INFO] consul: Registered fabio with check &{CheckID: Name: Args:[] DockerContainerID: Shell: Interval: Timeout: TTL:15s HTTP: Header:map[] Method: TCP: Status: Notes: TLSSkipVerify:false GRPC: GRPCUseTLS:false AliasNode: AliasService: DeregisterCriticalServiceAfter:15s}
2019/06/14 09:20:12 [INFO] consul: Registered fabio with check &{CheckID: Name: Args:[] DockerContainerID: Shell: Interval:1s Timeout:3s TTL: HTTP:http://[192.168.86.187]:9998/health Header:map[] Method: TCP: Status: Notes: TLSSkipVerify:false GRPC: GRPCUseTLS:false AliasNode: AliasService: DeregisterCriticalServiceAfter:90m}
2019/06/14 09:21:42 [INFO] consul: Registered fabio as "fabio"
2019/06/14 09:21:42 [INFO] consul: Registered fabio with id "fabio-pires-ipro.lan-9998"
2019/06/14 09:21:42 [INFO] consul: Registered fabio with address "192.168.86.187"
2019/06/14 09:21:42 [INFO] consul: Registered fabio with tags ""
2019/06/14 09:21:42 [INFO] consul: Registered fabio with check &{CheckID: Name: Args:[] DockerContainerID: Shell: Interval: Timeout: TTL:15s HTTP: Header:map[] Method: TCP: Status: Notes: TLSSkipVerify:false GRPC: GRPCUseTLS:false AliasNode: AliasService: DeregisterCriticalServiceAfter:15s}
2019/06/14 09:21:42 [INFO] consul: Registered fabio with check &{CheckID: Name: Args:[] DockerContainerID: Shell: Interval:1s Timeout:3s TTL: HTTP:http://[192.168.86.187]:9998/health Header:map[] Method: TCP: Status: Notes: TLSSkipVerify:false GRPC: GRPCUseTLS:false AliasNode: AliasService: DeregisterCriticalServiceAfter:90m}
^C2019/06/14 09:22:06 [INFO] Caught SIGINT. Exiting
2019/06/14 09:22:06 [INFO] consul: Deregistering "fabio"
2019/06/14 09:22:06 [INFO] Down
```